# FORGE — Walkthrough su dati reali AMZN 1D

Questo notebook percorre **l'intero pipeline FORGE** su 5 anni di candele giornaliere di Amazon (Jan 2021 → Jun 2026).

```
Candele OHLCV
  └─ build_features()      Feature engineering (EMA, RSI, BB, ATR, MACD, …)
       └─ forge()           Pipeline completa M0 → M1 → M2 → M3
            ├─ M0  MarketContext      — regime di mercato (bull/bear/…)
            ├─ M1  EventDiscovery     — segnali atomici consistenti
            ├─ M2  AlphaDiscovery     — contratti con direzione + holding period
            └─ M3  RuleDiscovery      — validazione IS + walk-forward OOS
```

**Output finale:** regole di trading statisticamente robuste (PARTIAL-EDGE / EDGE) già backtestabili.

## 1 — Setup

In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from pathlib import Path
from collections import Counter

from forgedge import build_features, forge, forge_preset

DATA = Path("data/AMZN_1D.csv")

## 2 — Caricamento e pulizia dati

Il file CSV viene da Investing.com (formato standard: Date, Price, Open, High, Low, Vol., Change %).
Le colonne vengono rinominate e il volume viene convertito da formato `"31.50M"` a float.

In [2]:
raw = pd.read_csv(DATA)

# normalise column names: lower, no dots/spaces
raw.columns = [c.strip().lower().replace(".", "").replace(" ", "_") for c in raw.columns]

# rename to OHLCV convention
raw = raw.rename(columns={"price": "close", "vol": "volume", "change_%": "chg_pct"})

# parse date and sort ascending
raw["open_dt"] = pd.to_datetime(raw["date"], format="%m/%d/%Y")
raw = raw.sort_values("open_dt").reset_index(drop=True)

# numeric price columns
for col in ["open", "high", "low", "close"]:
    raw[col] = pd.to_numeric(
        raw[col].astype(str).str.replace(",", ""), errors="coerce"
    )

# volume: "31.50M" → 31_500_000 ; "1.2B" → 1_200_000_000
raw["volume"] = (
    raw["volume"]
    .astype(str)
    .str.replace("M", "e6")
    .str.replace("B", "e9")
)
raw["volume"] = pd.to_numeric(raw["volume"], errors="coerce").fillna(0)

# integer unix-ms timestamp required by build_features
raw["open_time"] = raw["open_dt"].astype("datetime64[ms]").astype("int64")

candles = raw[["open_time", "open_dt", "open", "high", "low", "close", "volume"]]

print(f"Barre: {len(candles)}")
print(f"Periodo: {candles['open_dt'].iloc[0].date()} → {candles['open_dt'].iloc[-1].date()}")
candles.tail(3)

Barre: 1378
Periodo: 2021-01-04 → 2026-06-30


,open_time,open_dt,open,high,low,close,volume
1375,1782432000000,2026-06-26,227.21,233.90,226.13,232.69,248370000.0
1376,1782691200000,2026-06-29,234.22,249.71,233.80,240.14,77620000.0
1377,1782777600000,2026-06-30,237.50,241.53,237.57,238.71,31500000.0


## 3 — Feature engineering

`build_features()` calcola indicatori tecnici sulle candele e restituisce la **KPI Table**: la struttura che serve a `forge()`. Ogni indicatore è descritto da un blocco di configurazione.

In [3]:
indicator_config = {
    "ema": {
        "enabled": True,
        "params": {"periods": [9, 50], "columns": ["close"]},
    },
    "rsi": {
        "enabled": True,
        "params": {"periods": [14, 25], "columns": ["close"]},
    },
    "bollinger_bands": {
        "enabled": True,
        "params": {"periods": [20], "columns": ["close"]},
    },
    "atr": {
        "enabled": True,
        "params": {"periods": [14], "columns": ["close"]},
    },
    "macd": {
        "enabled": True,
        "params": {"fast": 12, "slow": 26, "signal": 9, "columns": ["close"]},
    },
}

kpi = build_features(candles, indicator_config, timestamp_col="open_time")

feature_cols = [c for c in kpi.columns if c not in candles.columns]
print(f"KPI Table: {kpi.shape[0]} righe × {kpi.shape[1]} colonne")
print(f"Feature generate: {len(feature_cols)}")
print("Esempi:", ", ".join(feature_cols[:10]), "…")

indicatore 'atr' non implementato, salto


indicatore 'macd' non implementato, salto


KPI Table: 1378 righe × 16 colonne
Feature generate: 9
Esempi: color, close_ema_09, close_ema_50, close_rsi_14, close_rsi_25, close_bb_mid_20, close_bb_upper_20, close_bb_lower_20, close_bb_width_20 …


## 4 — Configurazione preset

`forge_preset()` restituisce una tripletta di config pre-calibrate per il timeframe e l'asset.

| Preset | Filosofia | Copertura |
|--------|-----------|----------|
| **sniper** | solo segnali ad alta coerenza temporale | bassa |
| **balanced** | equilibrio qualità/copertura | media |
| **sweep** | esplora tutto, filtra poco | alta |
| **burst** | segnali ad alta frequenza di attivazione | media |

Per AMZN 1D usiamo **sweep** per massimizzare la copertura esplorativa.

In [4]:
event_cfg, alpha_cfg, rd_cfg = forge_preset(
    "sweep",
    timeframe="1D",
    asset="AMZN",
)

gp = event_cfg.gate_params
sc = rd_cfg.criteria
print(f"Gate: min_tpm={gp.min_tpm:.2f} ep/mese  max_dispersion={gp.max_dispersion}  mode='{gp.event_counting}'")
print(f"RD:   min_profit_factor={sc.min_profit_factor}  min_tpm={sc.min_tpm}")

Gate: min_tpm=0.30 ep/mese  max_dispersion=2.5  mode='episode'
RD:   min_profit_factor=2.0  min_tpm=0.3


## 5 — Esecuzione FORGE

Un'unica chiamata `forge()` esegue l'intero pipeline M0 → M3.

In [5]:
result = forge(
    kpi,
    ticker="AMZN",
    timeframe="1D",
    event_discovery_config=event_cfg,
    alpha_config=alpha_cfg,
    rule_discovery_config=rd_cfg,
    run_registry=False,
    progress=True,
)

print(f"\nM1 candidati  : {len(result.candidates)}")
print(f"M2 contratti  : {len(result.contracts)}")
print(f"M2 promossi   : {len(result.promoted)}")
print(f"M3 risposte   : {len(result.rule_responses)}")

[forge:AMZN +   0.0s] start — 1378 bars


[forge:AMZN +   0.0s] M0 Market Context — classifying regimes…


[forge:AMZN +   0.1s] M1 Event Discovery — mining event candidates…


[forge:AMZN +   6.5s] M1 Event Discovery — 2717 candidate(s)


[forge:AMZN +   6.5s] M2 Alpha Discovery — evaluating 2717 candidate(s)…


[forge:AMZN +  16.0s] M2 Alpha Discovery — 405/2717 promoted


[forge:AMZN +  16.0s] M3 Rule Discovery — backtesting 405 contract(s)…


[forge:AMZN] M3 Rule Discovery |····························| 1/405 (  0%)   0.3s

[forge:AMZN] M3 Rule Discovery |····························| 2/405 (  0%)   0.6s

[forge:AMZN] M3 Rule Discovery |····························| 3/405 (  1%)   0.7s

[forge:AMZN] M3 Rule Discovery |····························| 4/405 (  1%)   1.6s

[forge:AMZN] M3 Rule Discovery |····························| 5/405 (  1%)   2.2s

[forge:AMZN] M3 Rule Discovery |····························| 6/405 (  1%)   2.4s

[forge:AMZN] M3 Rule Discovery |····························| 7/405 (  2%)   2.9s

[forge:AMZN] M3 Rule Discovery |····························| 8/405 (  2%)   3.4s

[forge:AMZN] M3 Rule Discovery |····························| 9/405 (  2%)   3.6s

[forge:AMZN] M3 Rule Discovery |····························| 10/405 (  2%)   3.7s

[forge:AMZN] M3 Rule Discovery |····························| 11/405 (  3%)   3.9s

[forge:AMZN] M3 Rule Discovery |····························| 12/405 (  3%)   4.5s

[forge:AMZN] M3 Rule Discovery |····························| 13/405 (  3%)   5.2s

[forge:AMZN] M3 Rule Discovery |····························| 14/405 (  3%)   6.0s

[forge:AMZN] M3 Rule Discovery |█···························| 15/405 (  4%)   6.6s

[forge:AMZN] M3 Rule Discovery |█···························| 16/405 (  4%)   7.5s

[forge:AMZN] M3 Rule Discovery |█···························| 17/405 (  4%)   7.6s

[forge:AMZN] M3 Rule Discovery |█···························| 18/405 (  4%)   8.7s

[forge:AMZN] M3 Rule Discovery |█···························| 19/405 (  5%)   9.3s

[forge:AMZN] M3 Rule Discovery |█···························| 20/405 (  5%)   9.7s

[forge:AMZN] M3 Rule Discovery |█···························| 21/405 (  5%)  10.2s

[forge:AMZN] M3 Rule Discovery |█···························| 22/405 (  5%)  10.6s

[forge:AMZN] M3 Rule Discovery |█···························| 23/405 (  6%)  11.2s

[forge:AMZN] M3 Rule Discovery |█···························| 24/405 (  6%)  11.8s

[forge:AMZN] M3 Rule Discovery |█···························| 25/405 (  6%)  11.9s

[forge:AMZN] M3 Rule Discovery |█···························| 26/405 (  6%)  12.1s

[forge:AMZN] M3 Rule Discovery |█···························| 27/405 (  7%)  12.3s

[forge:AMZN] M3 Rule Discovery |█···························| 28/405 (  7%)  12.4s

[forge:AMZN] M3 Rule Discovery |██··························| 29/405 (  7%)  13.1s

[forge:AMZN] M3 Rule Discovery |██··························| 30/405 (  7%)  13.9s

[forge:AMZN] M3 Rule Discovery |██··························| 31/405 (  8%)  14.8s

[forge:AMZN] M3 Rule Discovery |██··························| 32/405 (  8%)  15.1s

[forge:AMZN] M3 Rule Discovery |██··························| 33/405 (  8%)  15.8s

[forge:AMZN] M3 Rule Discovery |██··························| 34/405 (  8%)  17.0s

[forge:AMZN] M3 Rule Discovery |██··························| 35/405 (  9%)  17.1s

[forge:AMZN] M3 Rule Discovery |██··························| 36/405 (  9%)  17.3s

[forge:AMZN] M3 Rule Discovery |██··························| 37/405 (  9%)  18.0s

[forge:AMZN] M3 Rule Discovery |██··························| 38/405 (  9%)  18.2s

[forge:AMZN] M3 Rule Discovery |██··························| 39/405 ( 10%)  18.4s

[forge:AMZN] M3 Rule Discovery |██··························| 40/405 ( 10%)  18.5s

[forge:AMZN] M3 Rule Discovery |██··························| 41/405 ( 10%)  19.0s

[forge:AMZN] M3 Rule Discovery |██··························| 42/405 ( 10%)  19.2s

[forge:AMZN] M3 Rule Discovery |██··························| 43/405 ( 11%)  19.3s

[forge:AMZN] M3 Rule Discovery |███·························| 44/405 ( 11%)  19.7s

[forge:AMZN] M3 Rule Discovery |███·························| 45/405 ( 11%)  19.8s

[forge:AMZN] M3 Rule Discovery |███·························| 46/405 ( 11%)  20.0s

[forge:AMZN] M3 Rule Discovery |███·························| 47/405 ( 12%)  20.1s

[forge:AMZN] M3 Rule Discovery |███·························| 48/405 ( 12%)  20.2s

[forge:AMZN] M3 Rule Discovery |███·························| 49/405 ( 12%)  20.9s

[forge:AMZN] M3 Rule Discovery |███·························| 50/405 ( 12%)  21.5s

[forge:AMZN] M3 Rule Discovery |███·························| 51/405 ( 13%)  21.6s

[forge:AMZN] M3 Rule Discovery |███·························| 52/405 ( 13%)  22.3s

[forge:AMZN] M3 Rule Discovery |███·························| 53/405 ( 13%)  23.0s

[forge:AMZN] M3 Rule Discovery |███·························| 54/405 ( 13%)  23.2s

[forge:AMZN] M3 Rule Discovery |███·························| 55/405 ( 14%)  23.8s

[forge:AMZN] M3 Rule Discovery |███·························| 56/405 ( 14%)  24.0s

[forge:AMZN] M3 Rule Discovery |███·························| 57/405 ( 14%)  24.1s

[forge:AMZN] M3 Rule Discovery |████························| 58/405 ( 14%)  24.3s

[forge:AMZN] M3 Rule Discovery |████························| 59/405 ( 15%)  24.9s

[forge:AMZN] M3 Rule Discovery |████························| 60/405 ( 15%)  25.0s

[forge:AMZN] M3 Rule Discovery |████························| 61/405 ( 15%)  25.5s

[forge:AMZN] M3 Rule Discovery |████························| 62/405 ( 15%)  25.7s

[forge:AMZN] M3 Rule Discovery |████························| 63/405 ( 16%)  25.7s

[forge:AMZN] M3 Rule Discovery |████························| 64/405 ( 16%)  25.8s

[forge:AMZN] M3 Rule Discovery |████························| 65/405 ( 16%)  26.4s

[forge:AMZN] M3 Rule Discovery |████························| 66/405 ( 16%)  26.5s

[forge:AMZN] M3 Rule Discovery |████························| 67/405 ( 17%)  26.6s

[forge:AMZN] M3 Rule Discovery |████························| 68/405 ( 17%)  26.7s

[forge:AMZN] M3 Rule Discovery |████························| 69/405 ( 17%)  26.8s

[forge:AMZN] M3 Rule Discovery |████························| 70/405 ( 17%)  26.9s

[forge:AMZN] M3 Rule Discovery |████························| 71/405 ( 18%)  27.1s

[forge:AMZN] M3 Rule Discovery |████························| 72/405 ( 18%)  27.2s

[forge:AMZN] M3 Rule Discovery |█████·······················| 73/405 ( 18%)  27.7s

[forge:AMZN] M3 Rule Discovery |█████·······················| 74/405 ( 18%)  27.8s

[forge:AMZN] M3 Rule Discovery |█████·······················| 75/405 ( 19%)  27.9s

[forge:AMZN] M3 Rule Discovery |█████·······················| 76/405 ( 19%)  28.1s

[forge:AMZN] M3 Rule Discovery |█████·······················| 77/405 ( 19%)  28.2s

[forge:AMZN] M3 Rule Discovery |█████·······················| 78/405 ( 19%)  28.3s

[forge:AMZN] M3 Rule Discovery |█████·······················| 79/405 ( 20%)  28.5s

[forge:AMZN] M3 Rule Discovery |█████·······················| 80/405 ( 20%)  28.6s

[forge:AMZN] M3 Rule Discovery |█████·······················| 81/405 ( 20%)  28.8s

[forge:AMZN] M3 Rule Discovery |█████·······················| 82/405 ( 20%)  29.4s

[forge:AMZN] M3 Rule Discovery |█████·······················| 83/405 ( 20%)  29.5s

[forge:AMZN] M3 Rule Discovery |█████·······················| 84/405 ( 21%)  29.6s

[forge:AMZN] M3 Rule Discovery |█████·······················| 85/405 ( 21%)  30.2s

[forge:AMZN] M3 Rule Discovery |█████·······················| 86/405 ( 21%)  30.4s

[forge:AMZN] M3 Rule Discovery |██████······················| 87/405 ( 21%)  30.5s

[forge:AMZN] M3 Rule Discovery |██████······················| 88/405 ( 22%)  31.1s

[forge:AMZN] M3 Rule Discovery |██████······················| 89/405 ( 22%)  31.2s

[forge:AMZN] M3 Rule Discovery |██████······················| 90/405 ( 22%)  31.3s

[forge:AMZN] M3 Rule Discovery |██████······················| 91/405 ( 22%)  31.5s

[forge:AMZN] M3 Rule Discovery |██████······················| 92/405 ( 23%)  31.6s

[forge:AMZN] M3 Rule Discovery |██████······················| 93/405 ( 23%)  31.7s

[forge:AMZN] M3 Rule Discovery |██████······················| 94/405 ( 23%)  31.9s

[forge:AMZN] M3 Rule Discovery |██████······················| 95/405 ( 23%)  32.0s

[forge:AMZN] M3 Rule Discovery |██████······················| 96/405 ( 24%)  32.1s

[forge:AMZN] M3 Rule Discovery |██████······················| 97/405 ( 24%)  32.2s

[forge:AMZN] M3 Rule Discovery |██████······················| 98/405 ( 24%)  32.3s

[forge:AMZN] M3 Rule Discovery |██████······················| 99/405 ( 24%)  32.4s

[forge:AMZN] M3 Rule Discovery |██████······················| 100/405 ( 25%)  32.5s

[forge:AMZN] M3 Rule Discovery |██████······················| 101/405 ( 25%)  32.6s

[forge:AMZN] M3 Rule Discovery |███████·····················| 102/405 ( 25%)  33.2s

[forge:AMZN] M3 Rule Discovery |███████·····················| 103/405 ( 25%)  33.3s

[forge:AMZN] M3 Rule Discovery |███████·····················| 104/405 ( 26%)  33.3s

[forge:AMZN] M3 Rule Discovery |███████·····················| 105/405 ( 26%)  33.4s

[forge:AMZN] M3 Rule Discovery |███████·····················| 106/405 ( 26%)  33.6s

[forge:AMZN] M3 Rule Discovery |███████·····················| 107/405 ( 26%)  33.7s

[forge:AMZN] M3 Rule Discovery |███████·····················| 108/405 ( 27%)  33.8s

[forge:AMZN] M3 Rule Discovery |███████·····················| 109/405 ( 27%)  33.9s

[forge:AMZN] M3 Rule Discovery |███████·····················| 110/405 ( 27%)  34.0s

[forge:AMZN] M3 Rule Discovery |███████·····················| 111/405 ( 27%)  34.1s

[forge:AMZN] M3 Rule Discovery |███████·····················| 112/405 ( 28%)  34.2s

[forge:AMZN] M3 Rule Discovery |███████·····················| 113/405 ( 28%)  34.4s

[forge:AMZN] M3 Rule Discovery |███████·····················| 114/405 ( 28%)  35.0s

[forge:AMZN] M3 Rule Discovery |███████·····················| 115/405 ( 28%)  35.1s

[forge:AMZN] M3 Rule Discovery |████████····················| 116/405 ( 29%)  35.5s

[forge:AMZN] M3 Rule Discovery |████████····················| 117/405 ( 29%)  36.1s

[forge:AMZN] M3 Rule Discovery |████████····················| 118/405 ( 29%)  36.8s

[forge:AMZN] M3 Rule Discovery |████████····················| 119/405 ( 29%)  36.8s

[forge:AMZN] M3 Rule Discovery |████████····················| 120/405 ( 30%)  37.0s

[forge:AMZN] M3 Rule Discovery |████████····················| 121/405 ( 30%)  37.6s

[forge:AMZN] M3 Rule Discovery |████████····················| 122/405 ( 30%)  38.2s

[forge:AMZN] M3 Rule Discovery |████████····················| 123/405 ( 30%)  38.8s

[forge:AMZN] M3 Rule Discovery |████████····················| 124/405 ( 31%)  39.0s

[forge:AMZN] M3 Rule Discovery |████████····················| 125/405 ( 31%)  39.1s

[forge:AMZN] M3 Rule Discovery |████████····················| 126/405 ( 31%)  39.2s

[forge:AMZN] M3 Rule Discovery |████████····················| 127/405 ( 31%)  39.3s

[forge:AMZN] M3 Rule Discovery |████████····················| 128/405 ( 32%)  39.9s

[forge:AMZN] M3 Rule Discovery |████████····················| 129/405 ( 32%)  40.1s

[forge:AMZN] M3 Rule Discovery |████████····················| 130/405 ( 32%)  40.2s

[forge:AMZN] M3 Rule Discovery |█████████···················| 131/405 ( 32%)  40.3s

[forge:AMZN] M3 Rule Discovery |█████████···················| 132/405 ( 33%)  40.9s

[forge:AMZN] M3 Rule Discovery |█████████···················| 133/405 ( 33%)  41.1s

[forge:AMZN] M3 Rule Discovery |█████████···················| 134/405 ( 33%)  41.2s

[forge:AMZN] M3 Rule Discovery |█████████···················| 135/405 ( 33%)  41.6s

[forge:AMZN] M3 Rule Discovery |█████████···················| 136/405 ( 34%)  42.1s

[forge:AMZN] M3 Rule Discovery |█████████···················| 137/405 ( 34%)  42.5s

[forge:AMZN] M3 Rule Discovery |█████████···················| 138/405 ( 34%)  42.9s

[forge:AMZN] M3 Rule Discovery |█████████···················| 139/405 ( 34%)  43.0s

[forge:AMZN] M3 Rule Discovery |█████████···················| 140/405 ( 35%)  43.1s

[forge:AMZN] M3 Rule Discovery |█████████···················| 141/405 ( 35%)  43.6s

[forge:AMZN] M3 Rule Discovery |█████████···················| 142/405 ( 35%)  43.7s

[forge:AMZN] M3 Rule Discovery |█████████···················| 143/405 ( 35%)  43.8s

[forge:AMZN] M3 Rule Discovery |█████████···················| 144/405 ( 36%)  44.5s

[forge:AMZN] M3 Rule Discovery |██████████··················| 145/405 ( 36%)  44.6s

[forge:AMZN] M3 Rule Discovery |██████████··················| 146/405 ( 36%)  45.1s

[forge:AMZN] M3 Rule Discovery |██████████··················| 147/405 ( 36%)  45.7s

[forge:AMZN] M3 Rule Discovery |██████████··················| 148/405 ( 37%)  45.9s

[forge:AMZN] M3 Rule Discovery |██████████··················| 149/405 ( 37%)  46.5s

[forge:AMZN] M3 Rule Discovery |██████████··················| 150/405 ( 37%)  47.1s

[forge:AMZN] M3 Rule Discovery |██████████··················| 151/405 ( 37%)  47.3s

[forge:AMZN] M3 Rule Discovery |██████████··················| 152/405 ( 38%)  47.5s

[forge:AMZN] M3 Rule Discovery |██████████··················| 153/405 ( 38%)  47.6s

[forge:AMZN] M3 Rule Discovery |██████████··················| 154/405 ( 38%)  47.7s

[forge:AMZN] M3 Rule Discovery |██████████··················| 155/405 ( 38%)  47.9s

[forge:AMZN] M3 Rule Discovery |██████████··················| 156/405 ( 39%)  48.5s

[forge:AMZN] M3 Rule Discovery |██████████··················| 157/405 ( 39%)  49.3s

[forge:AMZN] M3 Rule Discovery |██████████··················| 158/405 ( 39%)  49.5s

[forge:AMZN] M3 Rule Discovery |██████████··················| 159/405 ( 39%)  49.6s

[forge:AMZN] M3 Rule Discovery |███████████·················| 160/405 ( 40%)  50.3s

[forge:AMZN] M3 Rule Discovery |███████████·················| 161/405 ( 40%)  50.4s

[forge:AMZN] M3 Rule Discovery |███████████·················| 162/405 ( 40%)  50.6s

[forge:AMZN] M3 Rule Discovery |███████████·················| 163/405 ( 40%)  50.7s

[forge:AMZN] M3 Rule Discovery |███████████·················| 164/405 ( 40%)  51.4s

[forge:AMZN] M3 Rule Discovery |███████████·················| 165/405 ( 41%)  51.7s

[forge:AMZN] M3 Rule Discovery |███████████·················| 166/405 ( 41%)  51.9s

[forge:AMZN] M3 Rule Discovery |███████████·················| 167/405 ( 41%)  52.0s

[forge:AMZN] M3 Rule Discovery |███████████·················| 168/405 ( 41%)  52.2s

[forge:AMZN] M3 Rule Discovery |███████████·················| 169/405 ( 42%)  52.3s

[forge:AMZN] M3 Rule Discovery |███████████·················| 170/405 ( 42%)  52.4s

[forge:AMZN] M3 Rule Discovery |███████████·················| 171/405 ( 42%)  52.6s

[forge:AMZN] M3 Rule Discovery |███████████·················| 172/405 ( 42%)  53.3s

[forge:AMZN] M3 Rule Discovery |███████████·················| 173/405 ( 43%)  53.9s

[forge:AMZN] M3 Rule Discovery |████████████················| 174/405 ( 43%)  54.5s

[forge:AMZN] M3 Rule Discovery |████████████················| 175/405 ( 43%)  55.0s

[forge:AMZN] M3 Rule Discovery |████████████················| 176/405 ( 43%)  55.6s

[forge:AMZN] M3 Rule Discovery |████████████················| 177/405 ( 44%)  56.3s

[forge:AMZN] M3 Rule Discovery |████████████················| 178/405 ( 44%)  56.4s

[forge:AMZN] M3 Rule Discovery |████████████················| 179/405 ( 44%)  56.5s

[forge:AMZN] M3 Rule Discovery |████████████················| 180/405 ( 44%)  56.6s

[forge:AMZN] M3 Rule Discovery |████████████················| 181/405 ( 45%)  57.2s

[forge:AMZN] M3 Rule Discovery |████████████················| 182/405 ( 45%)  57.2s

[forge:AMZN] M3 Rule Discovery |████████████················| 183/405 ( 45%)  57.3s

[forge:AMZN] M3 Rule Discovery |████████████················| 184/405 ( 45%)  57.4s

[forge:AMZN] M3 Rule Discovery |████████████················| 185/405 ( 46%)  57.5s

[forge:AMZN] M3 Rule Discovery |████████████················| 186/405 ( 46%)  57.7s

[forge:AMZN] M3 Rule Discovery |████████████················| 187/405 ( 46%)  58.2s

[forge:AMZN] M3 Rule Discovery |████████████················| 188/405 ( 46%)  58.7s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 189/405 ( 47%)  58.8s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 190/405 ( 47%)  59.4s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 191/405 ( 47%)  59.9s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 192/405 ( 47%)  60.3s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 193/405 ( 48%)  60.9s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 194/405 ( 48%)  61.1s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 195/405 ( 48%)  61.7s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 196/405 ( 48%)  61.8s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 197/405 ( 49%)  61.9s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 198/405 ( 49%)  62.0s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 199/405 ( 49%)  62.0s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 200/405 ( 49%)  62.6s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 201/405 ( 50%)  62.8s

[forge:AMZN] M3 Rule Discovery |█████████████···············| 202/405 ( 50%)  63.4s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 203/405 ( 50%)  63.5s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 204/405 ( 50%)  64.0s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 205/405 ( 51%)  64.6s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 206/405 ( 51%)  65.2s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 207/405 ( 51%)  65.8s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 208/405 ( 51%)  66.0s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 209/405 ( 52%)  66.4s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 210/405 ( 52%)  66.5s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 211/405 ( 52%)  67.1s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 212/405 ( 52%)  67.7s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 213/405 ( 53%)  68.4s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 214/405 ( 53%)  69.1s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 215/405 ( 53%)  69.6s

[forge:AMZN] M3 Rule Discovery |██████████████··············| 216/405 ( 53%)  70.3s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 217/405 ( 54%)  70.3s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 218/405 ( 54%)  71.0s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 219/405 ( 54%)  71.7s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 220/405 ( 54%)  72.2s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 221/405 ( 55%)  72.5s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 222/405 ( 55%)  73.2s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 223/405 ( 55%)  73.7s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 224/405 ( 55%)  74.3s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 225/405 ( 56%)  74.9s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 226/405 ( 56%)  75.5s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 227/405 ( 56%)  76.1s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 228/405 ( 56%)  76.6s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 229/405 ( 57%)  76.8s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 230/405 ( 57%)  76.9s

[forge:AMZN] M3 Rule Discovery |███████████████·············| 231/405 ( 57%)  77.5s

[forge:AMZN] M3 Rule Discovery |████████████████············| 232/405 ( 57%)  77.6s

[forge:AMZN] M3 Rule Discovery |████████████████············| 233/405 ( 58%)  78.3s

[forge:AMZN] M3 Rule Discovery |████████████████············| 234/405 ( 58%)  78.8s

[forge:AMZN] M3 Rule Discovery |████████████████············| 235/405 ( 58%)  79.4s

[forge:AMZN] M3 Rule Discovery |████████████████············| 236/405 ( 58%)  79.5s

[forge:AMZN] M3 Rule Discovery |████████████████············| 237/405 ( 59%)  79.6s

[forge:AMZN] M3 Rule Discovery |████████████████············| 238/405 ( 59%)  80.2s

[forge:AMZN] M3 Rule Discovery |████████████████············| 239/405 ( 59%)  80.8s

[forge:AMZN] M3 Rule Discovery |████████████████············| 240/405 ( 59%)  80.9s

[forge:AMZN] M3 Rule Discovery |████████████████············| 241/405 ( 60%)  81.5s

[forge:AMZN] M3 Rule Discovery |████████████████············| 242/405 ( 60%)  82.1s

[forge:AMZN] M3 Rule Discovery |████████████████············| 243/405 ( 60%)  82.7s

[forge:AMZN] M3 Rule Discovery |████████████████············| 244/405 ( 60%)  83.3s

[forge:AMZN] M3 Rule Discovery |████████████████············| 245/405 ( 60%)  83.8s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 246/405 ( 61%)  84.2s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 247/405 ( 61%)  84.9s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 248/405 ( 61%)  85.5s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 249/405 ( 61%)  85.8s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 250/405 ( 62%)  86.2s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 251/405 ( 62%)  86.3s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 252/405 ( 62%)  86.4s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 253/405 ( 62%)  87.1s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 254/405 ( 63%)  87.7s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 255/405 ( 63%)  88.1s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 256/405 ( 63%)  88.7s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 257/405 ( 63%)  89.1s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 258/405 ( 64%)  89.8s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 259/405 ( 64%)  89.9s

[forge:AMZN] M3 Rule Discovery |█████████████████···········| 260/405 ( 64%)  90.3s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 261/405 ( 64%)  90.3s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 262/405 ( 65%)  90.5s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 263/405 ( 65%)  91.4s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 264/405 ( 65%)  92.1s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 265/405 ( 65%)  92.5s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 266/405 ( 66%)  93.2s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 267/405 ( 66%)  94.0s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 268/405 ( 66%)  94.8s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 269/405 ( 66%)  95.3s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 270/405 ( 67%)  95.4s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 271/405 ( 67%)  95.5s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 272/405 ( 67%)  95.6s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 273/405 ( 67%)  96.2s

[forge:AMZN] M3 Rule Discovery |██████████████████··········| 274/405 ( 68%)  96.3s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 275/405 ( 68%)  96.6s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 276/405 ( 68%)  97.2s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 277/405 ( 68%)  97.4s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 278/405 ( 69%)  97.5s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 279/405 ( 69%)  97.7s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 280/405 ( 69%)  97.8s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 281/405 ( 69%)  98.0s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 282/405 ( 70%)  98.1s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 283/405 ( 70%)  98.2s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 284/405 ( 70%)  98.8s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 285/405 ( 70%)  99.4s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 286/405 ( 71%)  99.5s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 287/405 ( 71%)  99.7s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 288/405 ( 71%)  99.8s

[forge:AMZN] M3 Rule Discovery |███████████████████·········| 289/405 ( 71%)  99.9s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 290/405 ( 72%) 100.0s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 291/405 ( 72%) 100.1s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 292/405 ( 72%) 100.2s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 293/405 ( 72%) 100.3s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 294/405 ( 73%) 100.8s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 295/405 ( 73%) 101.4s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 296/405 ( 73%) 102.1s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 297/405 ( 73%) 102.3s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 298/405 ( 74%) 102.4s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 299/405 ( 74%) 102.5s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 300/405 ( 74%) 102.7s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 301/405 ( 74%) 103.2s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 302/405 ( 75%) 103.4s

[forge:AMZN] M3 Rule Discovery |████████████████████········| 303/405 ( 75%) 103.5s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 304/405 ( 75%) 103.7s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 305/405 ( 75%) 103.8s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 306/405 ( 76%) 104.0s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 307/405 ( 76%) 104.1s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 308/405 ( 76%) 104.2s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 309/405 ( 76%) 104.9s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 310/405 ( 77%) 105.4s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 311/405 ( 77%) 105.6s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 312/405 ( 77%) 105.7s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 313/405 ( 77%) 105.8s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 314/405 ( 78%) 106.0s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 315/405 ( 78%) 106.7s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 316/405 ( 78%) 106.8s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 317/405 ( 78%) 107.4s

[forge:AMZN] M3 Rule Discovery |█████████████████████·······| 318/405 ( 79%) 107.5s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 319/405 ( 79%) 107.6s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 320/405 ( 79%) 108.2s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 321/405 ( 79%) 108.8s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 322/405 ( 80%) 109.2s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 323/405 ( 80%) 109.8s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 324/405 ( 80%) 110.5s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 325/405 ( 80%) 110.9s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 326/405 ( 80%) 111.5s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 327/405 ( 81%) 112.1s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 328/405 ( 81%) 112.2s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 329/405 ( 81%) 112.8s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 330/405 ( 81%) 113.4s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 331/405 ( 82%) 113.6s

[forge:AMZN] M3 Rule Discovery |██████████████████████······| 332/405 ( 82%) 114.1s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 333/405 ( 82%) 114.3s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 334/405 ( 82%) 114.9s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 335/405 ( 83%) 115.0s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 336/405 ( 83%) 115.1s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 337/405 ( 83%) 115.6s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 338/405 ( 83%) 115.7s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 339/405 ( 84%) 115.8s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 340/405 ( 84%) 116.2s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 341/405 ( 84%) 116.3s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 342/405 ( 84%) 116.9s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 343/405 ( 85%) 116.9s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 344/405 ( 85%) 117.4s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 345/405 ( 85%) 118.0s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 346/405 ( 85%) 118.1s

[forge:AMZN] M3 Rule Discovery |███████████████████████·····| 347/405 ( 86%) 118.6s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 348/405 ( 86%) 118.7s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 349/405 ( 86%) 118.9s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 350/405 ( 86%) 119.0s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 351/405 ( 87%) 119.1s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 352/405 ( 87%) 119.2s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 353/405 ( 87%) 119.4s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 354/405 ( 87%) 119.5s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 355/405 ( 88%) 119.6s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 356/405 ( 88%) 120.2s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 357/405 ( 88%) 120.8s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 358/405 ( 88%) 121.4s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 359/405 ( 89%) 122.0s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 360/405 ( 89%) 122.1s

[forge:AMZN] M3 Rule Discovery |████████████████████████····| 361/405 ( 89%) 122.7s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 362/405 ( 89%) 123.3s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 363/405 ( 90%) 123.5s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 364/405 ( 90%) 123.6s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 365/405 ( 90%) 124.2s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 366/405 ( 90%) 124.8s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 367/405 ( 91%) 124.9s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 368/405 ( 91%) 125.0s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 369/405 ( 91%) 125.2s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 370/405 ( 91%) 125.3s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 371/405 ( 92%) 125.5s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 372/405 ( 92%) 126.0s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 373/405 ( 92%) 126.2s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 374/405 ( 92%) 126.3s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 375/405 ( 93%) 126.4s

[forge:AMZN] M3 Rule Discovery |█████████████████████████···| 376/405 ( 93%) 126.5s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 377/405 ( 93%) 126.7s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 378/405 ( 93%) 126.8s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 379/405 ( 94%) 127.3s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 380/405 ( 94%) 127.8s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 381/405 ( 94%) 127.9s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 382/405 ( 94%) 128.6s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 383/405 ( 95%) 128.7s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 384/405 ( 95%) 128.9s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 385/405 ( 95%) 129.5s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 386/405 ( 95%) 129.6s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 387/405 ( 96%) 129.8s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 388/405 ( 96%) 129.8s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 389/405 ( 96%) 130.4s

[forge:AMZN] M3 Rule Discovery |██████████████████████████··| 390/405 ( 96%) 131.0s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 391/405 ( 97%) 131.1s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 392/405 ( 97%) 131.2s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 393/405 ( 97%) 131.9s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 394/405 ( 97%) 132.0s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 395/405 ( 98%) 132.8s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 396/405 ( 98%) 132.9s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 397/405 ( 98%) 133.4s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 398/405 ( 98%) 133.5s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 399/405 ( 99%) 133.6s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 400/405 ( 99%) 134.2s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 401/405 ( 99%) 134.7s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 402/405 ( 99%) 134.7s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 403/405 (100%) 135.2s

[forge:AMZN] M3 Rule Discovery |███████████████████████████·| 404/405 (100%) 135.7s

[forge:AMZN] M3 Rule Discovery |████████████████████████████| 405/405 (100%) 135.8s

[forge:AMZN + 151.8s] M3 Rule Discovery — 38 tradeable rule(s)


[forge:AMZN + 151.8s] done



M1 candidati  : 2717
M2 contratti  : 2717
M2 promossi   : 405
M3 risposte   : 405


## 6 — M1: Event Discovery

M1 genera **candidati atomici** (espressioni soglia `feature < valore`) filtrati dalla **ConsistencyGate** in base a frequenza mensile (`min_tpm`) e regolarità temporale (`max_dispersion` — indice di dispersione di Poisson). I candidati che superano il gate vengono composti in **AND** per formare segnali multivariati.

In [6]:
candidates = result.candidates
passed = [c for c in candidates if c.consistency_gate.passed]

print(f"Candidati totali : {len(candidates)}")
print(f"Gate superato    : {len(passed)}")
print(f"Atomici (1 feat) : {sum(1 for c in candidates if len(c.components) == 1)}")
print(f"AND composti     : {sum(1 for c in candidates if len(c.components) > 1)}")

Candidati totali : 2717
Gate superato    : 2717
Atomici (1 feat) : 717
AND composti     : 2000


In [7]:
# Top 10 per episodi/mese
rows = []
for c in sorted(candidates, key=lambda x: x.activation_stats.mean_tpm, reverse=True)[:10]:
    s = c.activation_stats
    rows.append({
        "event_id":   c.event_id,
        "expression": c.expression,
        "tpm":        round(s.mean_tpm, 3),
        "dispersion": round(s.index_of_dispersion, 3),
        "n_ep":       s.n_activations,
        "gate":       "✓" if c.consistency_gate.passed else "✗",
    })

pd.DataFrame(rows)

,event_id,expression,tpm,dispersion,n_ep,gate
0,EVT-ratio_close_ema09_em-ZS-0238,zs_ratio_close_ema09_ema50_48 < -1,5.439,8.520,359,✓
1,EVT-color-ZS-0025,zs_color_168 > 1,5.273,4.432,348,✓
2,EVT-diffnorm_close_ema09-ZS-0266,zs_diffnorm_close_ema09_ema50_48 > 1,5.212,8.699,344,✓
3,EVT-diffnorm_close_ema09-ZS-0265,zs_diffnorm_close_ema09_ema50_48 < -1,5.136,8.410,339,✓
4,EVT-ratio_close_ema09_em-ZS-0239,zs_ratio_close_ema09_ema50_48 > 1,4.985,8.342,329,✓
5,EVT-close_rsi_25-ZS-0140,zs_close_rsi_25_48 > 1,4.955,7.467,327,✓
6,EVT-close_bb_width_20-ZS-0206,zs_close_bb_width_20_48 > 1,4.939,6.465,326,✓
7,EVT-diffnorm_close_ema09-ZS-0268,zs_diffnorm_close_ema09_ema50_96 > 1,4.864,9.875,321,✓
8,EVT-spread_close_ema50-ZS-0573,zs_spread_close_ema50_48 < -1,4.742,6.750,313,✓
9,EVT-ratio_close_rsi14_rs-ZS-0322,zs_ratio_close_rsi14_rsi25_48 < -1,4.667,4.473,308,✓


## 7 — M2: Alpha Discovery

Per ogni candidato M1 promosso, M2 ricerca il **contratto ottimale**:
- **direzione** (long / short)
- **holding period h\*** (da 1 a `max_holding_period` barre)
- **sell_pct**: percentuale di gain target

Vengono promossi solo i contratti con `lift` statisticamente significativo.

In [8]:
promoted = result.promoted

print(f"Contratti totali M2 : {len(result.contracts)}")
print(f"Promossi            : {len(promoted)}")
print(f"  long              : {sum(1 for c in promoted if c.direction == 'long')}")
print(f"  short             : {sum(1 for c in promoted if c.direction == 'short')}")

Contratti totali M2 : 2717
Promossi            : 405
  long              : 204
  short             : 201


In [9]:
# Top 10 contratti promossi per lift
rows = []
for c in sorted(promoted, key=lambda x: x.event_stats.lift, reverse=True)[:10]:
    dt = c.derived_target
    rows.append({
        "expression":  c.event_expression,
        "direction":   c.direction,
        "h*":          dt.holding_period_h,
        "sell_pct %":  round(dt.sell_pct * 100, 2),
        "lift":        round(c.event_stats.lift, 4),
        "activations": c.event_stats.n_activations,
    })

pd.DataFrame(rows)

,expression,direction,h*,sell_pct %,lift,activations
0,delta_close_rsi_14_1 > 4.46123 AND close_rsi_2...,long,7,7.70,0.6952,15
1,zs_color_48 < -1 AND pr_close_rsi_14_168 < 0.1...,long,3,3.58,0.6562,11
2,zs_color_48 < -1 AND zs_close_rsi_14_168 < -1,long,3,3.72,0.5909,12
3,close_rsi_14 < 39.2903 AND delta_ratio_close_r...,long,10,7.89,0.5735,17
4,close_rsi_14 < 36.3905 AND delta_bb_pct_b_clos...,long,2,2.18,0.5385,18
5,close_rsi_14 < 36.3905 AND delta_spread_close_...,long,1,1.25,0.5253,22
6,close_rsi_25 < 40.3848,long,10,7.94,0.5219,96
7,delta_close_rsi_14_12 < -24.9697,long,2,1.99,0.5188,24
8,close_rsi_25 < 39.2649,long,10,9.34,0.5150,76
9,pr_color_168 < 0.247024 AND bb_pct_b_close_20 ...,long,2,2.11,0.5125,16


## 8 — M3: Rule Discovery

M3 valida ogni contratto promosso con:
1. **Backtest IS** — profit factor, win rate, expectancy
2. **Walk-forward OOS** — 4 split espandenti; consistency e degradazione
3. **Validazione statistica** — Deflated Sharpe, t-test win rate

Verdetti:
- `EDGE` — robusto IS + OOS
- `PARTIAL-EDGE` — IS forte, OOS sufficiente ma non pieno
- `NON-EDGE` — non supera i criteri

In [10]:
verdicts = Counter(r.verdict for _, r in result.rule_responses)
print("Verdetti M3:")
for v, n in sorted(verdicts.items(), key=lambda x: -x[1]):
    bar = "█" * n
    print(f"  {v:<16} {n:>4}  {bar}")

Verdetti M3:
  NON-EDGE          367  ███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████
  PARTIAL-EDGE       38  ██████████████████████████████████████


## 9 — Tabella PARTIAL-EDGE

Tutti i segnali con verdetto `PARTIAL-EDGE`, ordinati per profit factor IS.

In [11]:
partial_edge = [
    (cc, r)
    for cc, r in result.rule_responses
    if r.verdict == "PARTIAL-EDGE"
]
partial_edge.sort(key=lambda x: x[1].in_sample_summary.profit_factor, reverse=True)

rows = []
for cc, r in partial_edge:
    iss = r.in_sample_summary
    wf  = r.walk_forward
    oos = wf.oos_summary if wf else None
    dt  = cc.derived_target
    rows.append({
        "expression":  cc.event_expression,
        "dir":         cc.direction,
        "h*":          dt.holding_period_h if dt else None,
        "IS PF":       round(iss.profit_factor, 2),
        "IS WR %":     round(iss.win_rate_pct * 100, 1),
        "IS trades":   iss.total_trades,
        "OOS PF":      round(oos.profit_factor, 2) if oos and oos.total_trades else None,
        "OOS WR %":    round(oos.win_rate_pct * 100, 1) if oos and oos.total_trades else None,
        "OOS trades":  oos.total_trades if oos else None,
        "consistency": f"{wf.n_profitable_splits}/{len(wf.splits)}" if wf else None,
    })

df_pe = pd.DataFrame(rows)
print(f"{len(df_pe)} segnali PARTIAL-EDGE")
df_pe

38 segnali PARTIAL-EDGE


,expression,dir,h*,IS PF,IS WR %,IS trades,OOS PF,OOS WR %,OOS trades,consistency
0,zs_color_168 > 1 AND delta_ratio_close_ema09_e...,short,5,15.26,95.2,21,1.21,76.2,21,3/4
1,delta_close_rsi_14_1 > 4.46123 AND close_rsi_2...,short,10,8.69,95.2,21,3.52,75.0,20,4/4
2,pr_color_168 > 0.761905 AND delta_ratio_close_...,short,7,7.19,75.0,20,1.55,68.4,19,2/4
3,zs_color_168 > 1 AND delta_close_rsi_25_6 < -6...,short,5,6.72,90.5,21,4.19,89.5,19,4/4
4,zs_color_168 > 1 AND delta_ratio_close_ema09_e...,short,7,6.51,80.0,25,1.69,71.4,28,3/4
5,delta_close_rsi_14_3 > 8.10394 AND zs_close_bb...,short,3,6.27,82.6,23,3.94,80.0,20,3/4
6,delta_close_rsi_14_1 < -4.52196 AND delta_clos...,short,5,5.21,77.8,27,3.52,81.2,16,3/4
7,zs_color_168 > 1 AND pr_ratio_close_rsi14_rsi2...,short,7,4.12,95.0,20,1.37,77.3,22,3/4
8,zs_color_168 > 1 AND delta_ratio_close_rsi14_r...,short,7,3.98,70.0,20,1.22,55.6,18,3/4
9,zs_color_96 > 1 AND delta_ratio_close_rsi14_rs...,short,10,3.97,63.2,19,1.79,62.5,16,3/4


## 10 — Dettaglio sui segnali con 100% consistency

I segnali profittevoli in **tutti gli split OOS** sono i candidati più robusti per il paper trading.

In [12]:
def print_signal_detail(cc, r, rank):
    iss = r.in_sample_summary
    wf  = r.walk_forward
    oos = wf.oos_summary if wf else None
    dt  = cc.derived_target
    sv  = r.statistical_validation
    wfv = wf.oos_validation if wf else None

    print(f"{'═'*72}")
    print(f"  #{rank}  {r.verdict}")
    print(f"  Regola:  {cc.event_expression}")
    if dt:
        print(f"  Setup:   {cc.direction.upper()}  h*={dt.holding_period_h} barre  target={dt.sell_pct*100:.2f}%")
    print()
    print(f"  ┌─ IN-SAMPLE ({iss.total_trades} trades) ")
    print(f"  │  PF={iss.profit_factor:.3f}  WR={iss.win_rate_pct*100:.1f}%  Expect={iss.expectancy*100:.2f}%  TPM={iss.tpm_mu:.2f}")
    if sv:
        print(f"  │  DSharpe={sv.deflated_sharpe:.2f}  WR t={sv.ttest_winrate_t:.2f}  "
              f"trials={sv.n_trials_tested}  stability={sv.temporal_stability}")

    if oos and oos.total_trades:
        consistency = f"{wf.n_profitable_splits}/{len(wf.splits)}"
        print(f"  ├─ OOS AGGREGATO ({oos.total_trades} trades, {consistency} split profittevoli)")
        print(f"  │  PF={oos.profit_factor:.3f}  WR={oos.win_rate_pct*100:.1f}%  Expect={oos.expectancy*100:.2f}%  TPM={oos.tpm_mu:.2f}")
        if wfv:
            print(f"  │  stability={wfv.temporal_stability}  PF½1={wfv.pf_first_half:.2f}  PF½2={wfv.pf_second_half:.2f}")

    if wf and wf.splits:
        print(f"  ├─ WALK-FORWARD SPLITS")
        for sp in wf.splits:
            ts = sp.test_summary
            ok = "✓" if ts.profit_factor > 1 else "✗"
            pf_str = f"{ts.profit_factor:.2f}" if ts.profit_factor < 9999 else "∞"
            print(f"  │  [{ok}] {sp.test_from[:10]}→{sp.test_to[:10]}  "
                  f"PF={pf_str:<6}  WR={ts.win_rate_pct*100:.0f}%  t={ts.total_trades}")

    print(f"  └{'─'*69}")
    print()

In [13]:
robust = [
    (cc, r)
    for cc, r in partial_edge
    if r.walk_forward
    and r.walk_forward.n_profitable_splits == len(r.walk_forward.splits)
    and r.walk_forward.n_profitable_splits > 0
]

print(f"Segnali con consistency 100%: {len(robust)}\n")
for i, (cc, r) in enumerate(robust, 1):
    print_signal_detail(cc, r, i)

Segnali con consistency 100%: 2

════════════════════════════════════════════════════════════════════════
  #1  PARTIAL-EDGE
  Regola:  delta_close_rsi_14_1 > 4.46123 AND close_rsi_25 > 61.0731
  Setup:   SHORT  h*=10 barre  target=3.49%

  ┌─ IN-SAMPLE (21 trades) 
  │  PF=8.689  WR=95.2%  Expect=1.76%  TPM=0.32
  │  DSharpe=4.41  WR t=11.36  trials=75  stability=WARN
  ├─ OOS AGGREGATO (20 trades, 4/4 split profittevoli)
  │  PF=3.515  WR=75.0%  Expect=1.05%  TPM=0.33
  │  stability=WARN  PF½1=1.09  PF½2=15.62
  ├─ WALK-FORWARD SPLITS
  │  [✓] 2021-07-01→2022-10-01  PF=18.32   WR=50%  t=2
  │  [✓] 2022-10-01→2024-01-01  PF=1.23    WR=86%  t=7
  │  [✓] 2024-01-01→2025-04-01  PF=4.50    WR=57%  t=7
  │  [✓] 2025-04-01→2026-07-01  PF=∞       WR=100%  t=4
  └─────────────────────────────────────────────────────────────────────

════════════════════════════════════════════════════════════════════════
  #2  PARTIAL-EDGE
  Regola:  zs_color_168 > 1 AND delta_close_rsi_25_6 < -6.39394
  Setu

## 11 — Interpretazione

### Segnali chiave trovati da FORGE su AMZN 1D

**`zs_color_168 > 1 AND delta_close_rsi_25_6 < -6.39` — SHORT h\*=5**
- `zs_color_168 > 1`: momentum di mercato ribassista su 168 barre (le barre negative sono sopra la media z-score)
- `delta_close_rsi_25_6 < -6.39`: RSI25 sceso di almeno 6 punti in 6 sedute → segnale di accelerazione short
- **OOS**: PF ≈ 4.2, WR 89%, 100% consistency → stabile nel tempo

**`delta_close_rsi_14_1 > 4.46 AND close_rsi_25 > 61.07` — SHORT h\*=10**
- RSI14 spike giornaliero ≥ 4.5 punti mentre RSI25 è in zona ipercomprata (>61)
- Divergenza momentum breve / lungo termine → alta probabilità di correzione
- **OOS**: PF ≈ 3.5, WR 75%, 100% consistency

### Note sul verdetto PARTIAL-EDGE
- `PARTIAL-EDGE` ≠ segnale invalido: indica che l'edge c'è ma non ha ancora la significatività statistica piena
- Passo consigliato: paper trading 3-6 mesi, poi rivalutare con `EDGE` se il PF OOS live ≥ threshold
- Monitorare il **decay** del PF: se il PF OOS scende sotto 1.5 in live, il regime è cambiato

---

## Appendice — Struttura di `ForgeResult`

```python
result = forge(kpi, ...)

result.enriched          # KPI Table dopo MarketContext (+ colonna 'regime')
result.candidates        # List[EventCandidate]  — tutti i candidati M1
result.contracts         # List[AlphaContract]   — tutti i contratti M2
result.promoted          # List[AlphaContract]   — solo quelli promossi
result.rule_responses    # List[Tuple[AlphaContract, RuleDiscoveryResponse]]

# Per ogni risposta M3:
cc, r = result.rule_responses[0]
r.verdict                       # "EDGE" / "PARTIAL-EDGE" / "NON-EDGE"
r.in_sample_summary.profit_factor
r.in_sample_summary.win_rate_pct  # fraction [0, 1]
r.in_sample_summary.total_trades
r.in_sample_summary.expectancy
r.walk_forward.oos_summary        # BacktestSummary aggregato OOS
r.walk_forward.consistency        # frazione split profittevoli
r.walk_forward.splits             # List[WalkForwardSplit]
r.statistical_validation          # DSharpe, t-stat, temporal_stability

# Dal contratto M2:
cc.event_expression               # regola leggibile
cc.direction                      # "long" / "short"
cc.derived_target.holding_period_h
cc.derived_target.sell_pct
cc.event_stats.lift
cc.event_stats.n_activations
```